elfshaker as a C++ library

daminetreg opened the pull request:
This provides support via our custom [corrosion-rs](https://github.com/tipi-build/corrosion/) of elshaker as a library.

This provide all functions that are really in use in tipi as direct C++ calls with the same signature as we are used to in tipi::cache::elfshaker in the nxxm-src codebase.

See : 
https://github.com/tipi-build/elfshaker/blob/1c1145bec7d207105e6253516bcb9ca11143295e/src/lib.rs#L53-L105

This is a work-in-progress, the main remaining parts are : 
- [ ] adapting CI to use cmake exclusively (instead of cargo)
- [ ] moving files back to their original location by playing with the namespaces use since I reinstated the lib type instead of only staticlib which doesn't auto provide the library to the app ( so that we can continue to benefit from official elfshaker patching)

## Requires review and merge of 

- [ ] https://github.com/tipi-build/corrosion/pull/2
   - [ ] Providing patch to upstream
- [ ] https://github.com/cpp-pre/file/pull/3
- [ ] https://github.com/tipi-build/corrosion/pull/2

daminetreg committed:
:gear: CMakeList to link to elfshaker + make the elfshaker code a library as well that contains elfshaker binary utilities and adapt all modules to be public.

daminetreg committed:
:wrench: differently named executable and library elfshaker (for cxxbridge disambiguation) which can reuse library project modules.

The type staticlib doesn't produce a library that main.rs can reuse, while the type lib does.

daminetreg committed:
:gear: fully working cxxbridge to bring elfshaker functions as direct C++ api.

daminetreg committed:
:wink: more elfshaker C++ bridge and elfshaker repository with separate worktree support

daminetreg committed:
:wink: pack and status as direct C++ function with the ability to support separate worktree for elfshaker

daminetreg committed:
:mag: Full Rust + C++ test suite run via CMake and passing

This contains fixes and refactorig of duplicated code because of the newly
introducted do_[extract,pack...] functions.

daminetreg committed:
:gear: add corrosion as submodule as our official and public cmake-tipi-provider cannot handle these kinds of cmake modules repositories.

daminetreg committed:
:gear: suppport building the bridge from cmake-tipi-provider cached dependencies

daminetreg committed:
:barber: Now the `elfshaker` namespace

daminetreg committed:
:gear: working installation of elfshaker as a library

daminetreg committed:
:wrench: remove typo ( this was a copy paste kerfuffle )

daminetreg committed:
:wrench: don't export the rust library only the resulting bridge which

daminetreg committed:
:gear: updated to a newer version of corrosion that supports better exporting rust libraries (still requires an helping hand)

daminetreg committed:
:gear: aadding a CMake based CI workflow.

daminetreg committed:
:lipstick: syntax fixes to workflow Yaml

daminetreg committed:
:wrench: fix Ubuntu 16.04 builds by authorizing to use a node that works on Ubuntu 16.05 glibc

daminetreg committed:
:gear: remove dependency to gcc_s

Co-authored-by: Yannic Staudt <yannic@tipi.build>

daminetreg committed:
:new: Support full static linking via Corrosion builds

pysco68 committed:
:biohazard: WIP win64 build

pysco68 committed:
Revert ":wrench: cleaned up the environment variable injection for MSVCRT through corrosion-rs"

This reverts commit d66dce8cb7c39f6b541a05466cbb0b597a19a1cd.

# Conflicts:
#	NOTES_WIN64.md

pysco68 committed:
:wrench: cleaned up the environment variable injection for MSVCRT through corrosion-rs

daminetreg committed:
:gear: Ensure compilation is possible on Windows ARM

daminetreg committed:
:wrench: Fix path provided to elfshaker to be compatible with expecation of linux-style path avoiding panicking when removing repo prefix workdir.

daminetreg committed:
Merge branch 'feature/elfshaker-as-a-library-win64' of https://github.com/tipi-build/elfshaker into feature/elfshaker-as-a-library-win64

daminetreg committed:
:mag: incorporated a Corrosion fixe which runs cargo test as build system target to get the cargo test to run in Visual Studio Project contexts.

daminetreg committed:
:wrench: fix symlink storage, refactoring for separate worktree did mistakenly remove the base_dir out of symlinks targets from base_dir after this was already done in cases it was necessary

daminetreg committed:
:wrench: fix is_elfshaker_data_path checks to avoid the extra cost of filtering elfshaker_data/ on store or status.

daminetreg committed:
:wrench: don't filter symlinks to elfshaker worktree files in status detection.

This loop was misbehaving resulting in successful tests, it started to be
visible by the generalization of the strip_prefix(&base_dir) calls.

Before it basically always had the symlink target be inserted as relative
files, while the normalised path woud be as "./link.txt" the symlink
detection loop will insert it as "link.txt" without the "./"
( see old code https://github.com/tipi-build/elfshaker/commit/9a16d2839c589e65942ceb267938bb6ef4a5c400#diff-1b69f2595600a273d22bfe017eab8acd328ce69024315578fd71d8934be06e72L104-R120 )

What led me to the realization of what this loop was : a way to remove any
symlinks from status detection which was there by mistake.

daminetreg committed:
:wrench: repo_worktree and repo_datadir as std::string::String on all platforms.

daminetreg committed:
:mag: on macOS the Error can be NotFound for some of the failed AtomicUpdates.

daminetreg committed:
Fix test_atomic_update_api supporting WouldBlock as valid error code in the test

c.f. https://github.com/elfshaker/elfshaker/commit/bf6adf065464f2040b868cae9a7981695a89178b

daminetreg commented:
Now all tests pass on windows, macOS and linux

## windows
```
3:       Finished `release` profile [optimized] target(s) in 32.67s
3:        Running `C:/.tipi/vK.w/40999a5-elfshaker.b/b91cd2e/bin/x64/Release/cargo/build\release\deps\cli-cd5674971e4c89a5.exe`
3:   
3:   running 4 tests
3:   test package_file_see_status ... ok
3:   test symlink_to_directory ... ok
3:   test package_file_see_status_symlink ... ok
3:   test run_loosen_and_repack ... ok
3:
3:   test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.44s
3:
4/4 Test #3: cargo-test_cli ....................   Passed   36.37 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =  36.44 sec

C:\Users\daminetreg\workspace\elfshaker>C:\tipi-for-arm-cmake-tipi-provider.exe . -t  vs-16-2019-win64-cxx17 -C Release -vv -n --test all  -- -A x64 -T host=x64 
```

## macOS
```
1: test repo::fs::tests::test_remove_empty_dirs_works ... ok
1: test repo::repository::tests::compute_entry_diff_finds_updates ... ok
1: test status::test::no_middle ... ok
1: test status::test::add_back ... ok
1: test repo::repository::tests::compute_entry_diff_finds_update_of_duplicated ... ok
1: test repo::repository::tests::building_loose_object_paths_works ... ok
1: test atomicfile::tests::test_atomic_update_api ... ok
1: test repo::algo::test::run_in_parallel_works ... ok
1: 
1: test result: ok. 33 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.22s
1: 
1: 
4/4 Test #1: cargo-test_elfshaker ..............   Passed    6.47 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =   6.48 sec
➜  elfshaker git:(feature/elfshaker-as-a-library) ✗ tipi . -t clang-cxx17 -un --test all
```

## linux
```
3/4 Test #1: cargo-test_elfshaker ..............   Passed    9.02 sec
2:     Finished `test` profile [unoptimized + debuginfo] target(s) in 10.24s
2:      Running `/usr/local/share/.tipi/vK.w/40999a5-elfshaker.b/d53dd39/bin/./cargo/build/debug/deps/elfshaker_executable-8b17821f0d549e59`
2: 
2: running 0 tests
2: 
2: test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
2: 
4/4 Test #2: cargo-test_elfshaker_executable ...   Passed   10.31 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =  10.31 sec
daminetreg@damien-ubuntu:~/workspace/elfshaker$ tipi . -t linux-cxx17 -un --test all
```

daminetreg committed:
:gear: remove deprecated needless artifact upload.

daminetreg committed:
:gear: Get tipi and rustup ready on all 3 platforms for the build.

daminetreg committed:
:gear: fix the reference to the default tipi docker

daminetreg committed:
:gear: clone submodules too to have corrosion pulled in.

daminetreg committed:
:gear: set the HOME to the tipi user.

daminetreg committed:
:gear: always create a draft release

daminetreg committed:
:wrench: fix finding rust enabled with rustup 1.28

daminetreg committed:
:gear: add permission to create releases.

daminetreg committed:
:gear: ensure connectivity to tipi build cache is provided and use bigger CI runner on linux to alleviate space issues.

daminetreg committed:
:gear: ensure tipi distro is installed in full for builds on macOS and Windows

daminetreg committed:
:wrench: build and link with msvc as the default rustc.

daminetreg committed:
:gear: put secrets as quoted strings

daminetreg committed:
:gear: debug tipi connect

daminetreg committed:
:gear: don't quote the secrets this is not related.

daminetreg committed:
:gear: further debug outputs

daminetreg committed:
:biohazard: Debug by printing secrets that will be revoked later

daminetreg committed:
:gear: fixed secrets value and names

daminetreg committed:
:gear: macos x86 build by default

daminetreg committed:
:gear: run tests as separate step.

daminetreg committed:
:wrench: fix windows binary path.

daminetreg committed:
:wrench: enable x86 toolchain on macOS ARM for rustup + fix macos executable path

daminetreg committed:
:gear: use tipi with full path to run test separately as Path variable is not updated.

daminetreg committed:
:gear: debug rust x86 toolchain activation on macOS ARM.

daminetreg committed:
:wrench: fix tipi path on windows and force a run in an x86 context for the rustup x86 toolchain enablement

daminetreg committed:
:gear: read the rustup settings again.

daminetreg committed:
:wrench: force default host and toolchain on macOS ARM.

daminetreg committed:
:barber: reduce to the minimum and fix elfshaker path.

daminetreg committed:
:wrench: remove the old CI worfklow and run CI only on PRs.

daminetreg committed:
Merge pull request #13 from tipi-build/feature/elfshaker-as-a-library-debugci

Debugging CI with cmake-re + rustup and Github Actions Nodejs deprecations

Lambourl commented on .github/workflows/ci-cmake.yaml:
```diff
@@ -61,26 +67,37 @@ jobs:
           asset_content_type: application/zip
+      - name: run tests 
+        run: |
+          C:\Users\runneradmin\AppData\Local\tipi\tipi.exe . -t vs-16-2019-win64-cxx17 -C Release --dont-upgrade --verbose --use-cmakelists --test all
```
tipi should be on the path, please delete this long path

daminetreg replied:
It's not because the install script set the PATH In the other context not in that subshell. The install is not systemwide by default on windows.

Lambourl commented on .github/workflows/ci-cmake.yaml:
```diff
@@ -125,4 +149,7 @@ jobs:
-
+      - name: run tests
+        run: |
+          export HOME="/home/tipi"
```
give it in env to avoid repeat your self

daminetreg replied:
I suspect the result might be different as we don't override the user here and when loading the shell there will be some .bashrc processing and it could overwite an env setting made more generally before. Therefore it's not clean but I will keep this in this iteration.

Lambourl commented on .tipi/distro.json:
```diff
@@ -0,0 +1,424 @@
+      }
+    }
+  ]
+}
```
Did we need this file ? can we not use the distro or a modified distro with env var ?

daminetreg replied:
We need this and we don't have this file anywhere public otherwise. This is to represent the custom distro we use now with new environments in tipi, I need to update it though.

I could store it in the distro repo but it's for our own purpose not general.

Lambourl commented on .vscode/settings.json:
```diff
@@ -0,0 +1,10 @@
+    "iostream": "cpp",
+    "ostream": "cpp"
+  }
+}
```
i thins you need to delete this file

Lambourl commented on .tipi/opts.linux-cxx17:
```diff
@@ -0,0 +1,3 @@
+include_guard(GLOBAL)
```
For the first time, we protect our opts. Is it necessary?

daminetreg replied:
No it's not

Lambourl commented on CMakeLists.txt:
```diff
@@ -0,0 +1,154 @@
+include(FetchContent)
+set(FETCHCONTENT_QUIET OFF CACHE BOOL "" FORCE)
+
+#FetchContent_Declare(
+#    Corrosion
+#    GIT_REPOSITORY https://github.com/tipi-build/corrosion.git
+#    GIT_TAG c6132b645d1a0061a5c29ccf6a1d3c6f90490829 # v0.5 patched to support test execution, see https://github.com/tipi-build/corrosion/pull/1
+#)
+#FetchContent_MakeAvailable(Corrosion)
+set(CORROSION_VERBOSE_OUTPUT ON)
+add_subdirectory(cmake/corrosion)
```
Why do you have remove fetchcontent and use submodule ?

daminetreg replied:
Let's explain you this in a private channel

Lambourl commented on CMakeLists.txt:
```diff
@@ -0,0 +1,154 @@
+
+set(include_install_dir "include")
+
+add_executable(test_cxxbridge tests/test_cxxbridge.cpp)
+target_link_libraries(test_cxxbridge PUBLIC elfshaker-cxxbridge cpp-pre_file::file Boost::filesystem Boost::included_unit_test_framework)
+if (CMAKE_SYSTEM_NAME MATCHES Darwin)
+    target_link_libraries(test_cxxbridge PUBLIC "-framework Foundation")
+endif()
+add_test(NAME test_cxxbridge COMMAND $<TARGET_FILE:test_cxxbridge>) 
```
I will prefer to have it behind a `BUILD_TESTING` condition

daminetreg replied:
Good point

daminetreg replied:
Done in [8eb7325](https://github.com/tipi-build/elfshaker/pull/11/commits/8eb73255d0fc508db8000374fcc28d0680334db3)

Lambourl reviewed (changes requested):
Some small question or change. But the rust code is super clean

daminetreg committed:
Merge branch 'main' of github.com:tipi-build/elfshaker into feature/elfshaker-as-a-library

# Conflicts:
#	.github/workflows/ci-cmake.yaml

daminetreg committed:
:lipstick: run cargofmt for clippy-check and remove unnecessary printlines.

daminetreg committed:
:lipstick: no clippy check as the action is deprecated and archived.

daminetreg committed:
:lipstick: remove output of cpus detected

daminetreg committed:
:wrench: use actual_file_path to perform is_symlink check on repostory::store()

daminetreg committed:
:gear: ensure the elfshaker executable release still look the same, even if we won't depend on it anymore.

daminetreg committed:
:gear: build and register cargo and cmake tests only when BUILD_TESTING=ON


Supporting https://cmake.org/cmake/help/latest/variable/BUILD_TESTING.html
to address @Lambourl review point

daminetreg committed:
:lipstick: remove IDE configuration

daminetreg committed:
:wrench: add renamed elfshaker.exe to zip

Commit:
:wrench: have elfshaker (als "as a lib") take into account env $VERBOSE=1

daminetreg committed:
:book: add notes on how to build the software since the default won't build the tests anymore.

pysco68 reviewed (approved):
Thanks for the great work 🚀 

The few leftover calls the println() will be fixed in a subsequent PR

Lambourl reviewed (approved):
LGTM 
Thank you very much for your excellent work 👍
